File: tools/pkg/virtdeploy/cloudinit.go
package virtdeploy

import (
	"fmt"
	"os"
	"os/exec"
	"path"
)

func buildCloudInitIso(ciConfig *CloudInitConfig, isoPath string) error {
	if ciConfig == nil {
		return fmt.Errorf("cloud init config is nil")
	}

	tempDir, err := os.MkdirTemp("", "my-temp-dir-")
	if err != nil {
		return fmt.Errorf("failed to create temp dir: %w", err)
	}
	defer os.RemoveAll(tempDir)

	err = ciConfig.writeToDir(tempDir)
	if err != nil {
		return fmt.Errorf("failed to write cloud init config to dir: %w", err)
	}

	err = exec.Command(
		"xorrisofs",
		"-o",
		isoPath,
		"-J",
		"-input-charset",
		"utf8",
		"-rational-rock",
		"-V",
		"CIDATA",
		tempDir,
	).Run()
	if err != nil {
		return fmt.Errorf("failed to create cloud init ISO: %w", err)
	}

	return nil
}

func (ci *CloudInitConfig) writeToDir(dir string) error {
	stat, err := os.Stat(dir)
	if err != nil {
		return fmt.Errorf("failed to stat dir '%s': %w", dir, err)
	}

	if !stat.IsDir() {
		return fmt.Errorf("path '%s' is not a directory", dir)
	}

	err = os.WriteFile(path.Join(dir, "user-data"), []byte(ci.Userdata), 0600)
	if err != nil {
		return fmt.Errorf("failed to write user-data: %w", err)
	}

	err = os.WriteFile(path.Join(dir, "meta-data"), []byte(ci.Metadata), 0600)
	if err != nil {
		return fmt.Errorf("failed to write meta-data: %w", err)
	}

	return nil

}

File: tools/virtdeploy/helpers/cloudinit.py
from pathlib import Path
import shutil
import subprocess
import tempfile
from typing import Optional

from virtdeploy.helpers.vmmetadata import CloudInitConfig


def build_cloud_init_iso(output_path: Path, config: CloudInitConfig) -> None:

    tempdir = Path(tempfile.mkdtemp())
    try:
        shutil.copy(config.userdata, tempdir / "user-data")
        shutil.copy(config.metadata, tempdir / "meta-data")
        _create_iso(output_path, tempdir)
    finally:
        shutil.rmtree(tempdir, ignore_errors=True)


def _create_iso(output_path: Path, tempdir: Path) -> None:
    # Using the same iso options as virt-install
    # https://github.com/virt-manager/virt-manager/blob/f901c3277768a30c92daccc066b01784dccc1a05/virtinst/install/installerinject.py#L62C64-L62C80
    cmd = [
        "xorrisofs",
        "-o",
        str(output_path),
        "-J",
        "-input-charset",
        "utf8",
        "-rational-rock",
        "-V",
        "CIDATA",
        str(tempdir),
    ]

    result = subprocess.run(cmd, capture_output=True, text=True)
    if result.returncode != 0:
        raise RuntimeError(f"Failed to create cloud-init ISO: {result.stderr}")
